README.md

[![Arduino CI](https://github.com/RobTillaart/Cozir/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/Cozir/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/Cozir/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/Cozir/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/Cozir/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/Cozir/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/Cozir.svg?maxAge=3600)](https://github.com/RobTillaart/Cozir/releases)


# Cozir

Arduino library for COZIR range CO2 sensors.


## Description

The COZIR library is **experimental** as not all functionality is tested.
The polling mode as used in the examples is tested in the past by DirtGambit.
**CO2meter.com** sponsored a **COZIR GC0034** to start hands on testing (2022-02).
This sensor does not support all commands, but as the pattern of the commands 
is similar the non-tested are expected to work as well.

This version of the library supports only the **Serial** interface. 
Preferred is a hardware Serial port to connect the sensor but software Serial 
does work too. 

The library (since 0.3.4) a separate class to parse the STREAMING data.
See COZIRParser below. 
The COZIR class is focussed on polling and sending commands.


#### Notes

- Read the datasheet before using this library. 
It helps to understand the working of the COZIR sensor.
- Be aware that not all microprocessors have a SoftwareSerial library or one 
limited in performance. Polling at 9600 baud can be pretty blocking.
- the hardware serial based examples needs to be run on a MEGA or a Teensy, 
at least a board with more than one Serial port.
- If the device is in **CZR_COMMAND** mode it does not respond too polling calls.
It needs to be set to **CZR_POLLING** mode.
- Not all COZIR devices support all calls of this library.


## Interface Cozir

Read the datasheet (again).


### Constructor and initialisation

- **COZIR(Stream \* str)** constructor, gets a serial stream as reference.
- **void init()** sets operatingMode to **CZR_POLLING**
- **bool isInitialized()** returns true if enough time has passed after the call to **init()** for the sensor.
The sensor needs a few seconds to get correct values.


### Operating mode

- **void setOperatingMode(uint8_t mode)** set the operating mode either to **CZR_COMMAND**, **CZR_POLLING** or **CZR_STREAMING**
- **uint8_t getOperatingMode()** returns the mode set, **CZR_STREAMING** is the factory default. 
Please note that **init()** sets the operating mode to **CZR_POLLING**.


### Core functions

The COZIR CO2 sensors all support:
- **uint32_t CO2()** returns the CO2 concentration in PPM (!! might need PPMfactor).
- **uint16_t getPPMFactor()** returns 1, 10, 100. 
Normally the value returned is 1 but one should check at the first read and when there is a big jump in values returned.
Also when time interval between reads is large it might be useful to check the PPM factor.

Some COZIR sensors also support:

- **float celsius()** returns temperature of the sensor.
- **float fahrenheit()** idem, 'wrapper' around **celsius()**
- **float humidity()** idem, 'wrapper' around **celsius()**
- **float light()** idem.


### Calibration

Read datasheet before using these functions:

- **uint16_t fineTuneZeroPoint(uint16_t v1, uint16_t v2)**
- **uint16_t calibrateFreshAir()** typically 400 PPM.
- **uint16_t calibrateNitrogen()**
- **uint16_t calibrateKnownGas(uint16_t value)** 


#### Calibration NOT Recommended 

Following 3 functions are **NOT RECOMMENDED** by the datasheet.
Feel free to uncomment in the code and use at your own risk.
Read datasheet before using these functions:

- **uint16_t calibrateManual(uint16_t value)**
- **uint16_t setSpanCalibrate(uint16_t value)**
- **uint16_t getSpanCalibrate()**


### DigiFilter

use with care, read datasheet before use.

| value | meaning                                            |
|:-----:|:---------------------------------------------------|
|   0   | Special, see datasheet                             |
|   1   | fast, shows every single sample, raw, can be noisy |
|  32   | default, good average                              |
| 255   | very slow, max smoothed                            |


- **void setDigiFilter(uint8_t value)** The larger the value the more smoothed the signal is.
Larger values also means that the output does not follow fast changes.
So depending on your needs you need to find an optimal value for the project.
It might even so that you alternate between smooth and fast or adapt depending on
the actual CO2 value.
- **uint8_t getDigiFilter()** returns set value. 


### Streaming MODE

Warning: hardware serial is needed / recommended to improve the capture of all output correctly. 

- **void setOutputFields(uint16_t fields)** Sets the fields in the output stream as a 16 bit mask. See table below.
- **void clearOutputFields()** clears all the fields.
- **uint16_t getOutputFields()** returns the 16 bit mask of set output fields.
- **bool inOutputFields(uint16_t field)** returns true if the field is set.
- **void getRecentFields()** After a call to getRecentFields() you must read and parse the serial stream yourself.
The internal buffer of this Class cannot handle the possible large output. Lines can be over 100 bytes long!

The fields must be set as a bit mask, the order of the fields in the output is undetermined. 
So one need to parse the output of the sensor carefully.

Note: NOT all sensors support all fields, check the datasheet of the sensor used.

| Field             | Value HEX | Value DEC | Notes           |
|:------------------|:---------:|:---------:|:----------------|
| CZR_LIGHT         |  0X2000   |    8192   |                 |
| CZR_HUMIDITY      |  0X1000   |    4096   |                 |
| CZR_FILTLED       |  0X0800   |    2048   |                 |
| CZR_RAWLED        |  0X0400   |    1024   |                 |
| CZR_MAXLED        |  0X0200   |     512   |                 |
| CZR_ZEROPOINT     |  0X0100   |     256   |                 |
| CZR_RAWTEMP       |  0X0080   |     128   |                 |
| CZR_FILTTEMP      |  0X0040   |      64   |                 |
| CZR_FILTLEDSIGNAL |  0X0020   |      32   |                 |
| CZR_RAWLEDSIGNAL  |  0X0010   |      16   |                 |
| CZR_SENSTEMP      |  0X0008   |       8   |                 |
| CZR_FILTCO2       |  0X0004   |       4   |                 |
| CZR_RAWCO2        |  0X0002   |       2   |                 |
| CZR_NONE          |  0X0001   |       1   | reset           |
|                   |           |           |                 |
| CZR_DEFAULT       |  0X0006   |       6   | factory default |
| CZR_HTC           |  0X1082   |    4226   | shortcut        |
| CZR_ALL           |  0X3FFE   |   16383   | debug           |


Default value is CZR_DEFAULT ==  6  == CZR_FILTCO2 + CZR_RAWCO2


### EEPROM

Note: not all COZIR devices support EEPROM.
Check datasheet for the capabilities of your device.

In 0.3.0 the EEPROM function have been replaced by specific accessor functions. 
Read datasheet about the addresses and their meaning. Use with care.

- **void setAutoCalibrationPreload(uint16_t value)**
- **uint16_t getAutoCalibrationPreload()**
- **void setAutoCalibrationInterval(uint16_t value)**
- **uint16_t getAutoCalibrationInterval()**
- **void setAutoCalibrationOn()**
- **void setAutoCalibrationOff()**
- **bool getAutoCalibration()**
- **void setAutoCalibrationBackgroundConcentration(uint16_t value)**
- **uint16_t getAutoCalibrationBackgroundConcentration()**
- **void setAmbientConcentration(uint16_t value)**
- **uint16_t getAmbientConcentration()**
- **void setBufferClearTime(uint16_t value)**
- **uint16_t getBufferClearTime()**


#### EEPROM addresses used by above functions.

Read datasheet for the details and defaults of sensor at hand.

| Name     | Address | Default | Notes    |
|:---------|:-------:|:-------:|:---------|
| AHHI     | 0x00    | ?       | reserved |
| ANLO     | 0x01    | ?       | reserved |
| ANSOURCE | 0x02    | ?       | reserved |
| ACINITHI | 0x03    | 87      |          |
| ACINITLO | 0x04    | 192     |          |
| ACHI     | 0x05    | 94      |          |
| ACLO     | 0x06    | 128     |          |
| ACONOFF  | 0x07    | 0       |          |
| ACPPMHI  | 0x08    | 1       |          |
| ACPPMLO  | 0x09    | 194     |          |
| AMBHI    | 0x0A    | 1       |          |
| AMBLO    | 0x0B    | 194     |          |
| BCHI     | 0x0C    | 0       |          |
| BCLO     | 0x0D    | 8       |          |


### Miscellaneous

- **void getVersionSerial()** requests version over serial. 
The user should read (and parse) the serial output as it can become large. 
Also the user must reset the operating mode either to **CZR_POLLING** or **CZR_STREAMING**
- **void getConfiguration()** requests configuration over serial. 
The user should read (and parse) the serial output as it can become large. 
Also the user must reset the operating mode either to **CZR_POLLING** or **CZR_STREAMING**


## Operation

See examples.


## Future

- improve documentation
  - COZIR Parser a separate readme?
- add examples
  - example COZIR with I2C display?
- COZIR I2C class for newer generation 
  ~ same functional interface
  - multiWire / pin a la PCF8574 lib


#### won't for now

- add a **setEEPROMFactoryDefault()**?
  - unknown if all sensors have same values
- build a Arduino COZIR simulator for testing.
  - add other sensors underneath?



----


## COZIRParser

Class to parse the output of a COZIR sensor in stream mode.


## Description

(added in 0.3.4, experimental)

The COZIRparser object has a main function called **nextChar(c)**
It needs to be called with all incoming characters from the sensor.

If a new field is found **nextChar()** returns the field identifier 
of the last parsed field (single char) to indicate its value has been 
updated, otherwise it returns 0.
The updated value can be accessed with one of the functions, see cozir.h file.

An example **Cozir_stream_parse.ino** is added to show how to use this class.

**NOTE:** to send commands e.g. outputField selection, to the sensor the COZIR
 class can be used (which is mostly focussed on polling access). 
Alternatively the user sends the low level commands just as a string over serial.
This latter method will save memory especially in a final version of a project. 

**NOTE:** The COZIRparser skips the output of the Y, \* and @ command.
These are configuration fields and therefore not part of the **stream mode** fields.
Furthermore not all fields these lines produce are understood.
So parsing these lines is left to the user for now. 

**NOTE:** The COZIRparser class does not check for missing characters,
the range of the fields recognized, or other errors. So the values
returned should be handled with care.


## Interface COZIRParser

Read the datasheet (again).


### Constructor and initialisation

- **C0ZIRParser()** constructor
- **void init()** resets all internal variables to 0 except PPM (set to 1)
- **void resetParser()** should / could be called if stream from sensor is
interrupted for long period, e.g. if there are missing characters. 
This should prevent the parser to overshoot a value due to concatenating the 
value of different (interrupted) measurements or FIELDs.
- **uint8_t nextChar(char c)** all characters coming from the sensor should 
be send to the parser by means of **nextChar()**. Default it will return 0
and the FIELD character is a field has been updated. 
Read Datasheet for the chars used.

The remainder of the interface are getters for the different fields.


## Future

- improve documentation
  - COZIR Parser a separate readme?
- support splitting output of Y and \* command.
  - separate parser COZIRParserY()
  - separate parser COZIRParserStar()
- add examples
  - examples for COZIRParser.




changelog.md

# Cozir Changelog


## 0.3.5 2022-02-25
- added capturing mode 'K' field in the COZIRParser class
- added all known fields in the COZIRParser class
- added sending commands in stream parser example to test.
- minor edits.


## 0.3.4 2022-02-22
- added COZIRParser class for streaming mode
- added streaming examples
- minor edits


## 0.3.3 2022-02-21
- update readme,
- update + add examples
- tested with GC0034
- changed return value of **setOperatingMode()** to bool.
- refactor


## 0.3.2 2021-12-14
- update library.json
- license, minor edits


## 0.3.1 2021-10-20
- update Arduino-CI,
- badges in readme.md


## 0.3.0 2021-08-08
- Major update - breaks interface (names mainly)
- add **isInitialized()**
- add **getOperatingMode()**
- add **getOutputFields()**
- add **inOutputFields()**
- add **kelvin()**
- add EEPROM functions
- class methods camelCase
- extend unit tests



## 0.2.6 2021-01-31
- fix #4 use Mode0 for versions and configuration

## 0.2.5 2020-12-26
- fix software Serial + version number (oops)

## 0.2.2 2020-12-17
- add Arduino-CI + unit tests

## 0.2.1 2020-06-05
- fix library.json

## 0.2.0 2020-03-30
- some refactor and own repo

## 0.1.06
- added support for HardwareSerial for MEGA
- removed support for NewSoftSerial
- stop pre 1.0 support)

## 0.1.05
- fixed bug uint16_t request() to uint32_t request() in .h file

## 0.1.04
- changed CO2 to support larger values (Rob T)

## 0.1.03
- added setOperatingMode

## 0.1.02
- added support Arduino 1.x

## 0.1.01
- initial version



cozir.cpp
//
//    FILE: Cozir.cpp
//  AUTHOR: DirtGambit & Rob Tillaart
// VERSION: 0.3.5
// PURPOSE: library for COZIR range of sensors for Arduino
//          Polling Mode + stream parser
//     URL: https://github.com/RobTillaart/Cozir
//          http://forum.arduino.cc/index.php?topic=91467.0
//
// HISTORY: see changelog.md
//
// READ DATASHEET BEFORE USE OF THIS LIB !
//


#include "cozir.h"


#define CZR_INIT_DELAY              1200
#define CZR_REQUEST_TIMEOUT         200


// EEPROM ADRESSES
// P 11-12 manual     WHICH
//
//      Name          Address         Default value/ notes
#define CZR_AHHI        0x00            // reserved
#define CZR_ANLO        0x01            // reserved
#define CZR_ANSOURCE    0x02            // reserved
#define CZR_ACINITHI    0x03            // 87
#define CZR_ACINITLO    0x04            // 192
#define CZR_ACHI        0x05            // 94
#define CZR_ACLO        0x06            // 128
#define CZR_ACONOFF     0x07            // 0
#define CZR_ACPPMHI     0x08            // 1
#define CZR_ACPPMLO     0x09            // 194
#define CZR_AMBHI       0x0A            // 1
#define CZR_AMBLO       0x0B            // 194
#define CZR_BCHI        0x0C            // 0
#define CZR_BCLO        0x0D            // 8



COZIR::COZIR(Stream * str)
{
  _ser = str;
  _buffer[0] = '\0';
}


void COZIR::init()
{
  // override default streaming (takes too much performance)
  setOperatingMode(CZR_POLLING);
  _initTimeStamp = millis();
  // delay for initialization is kept until next major release.
  //    timestamp + isInitialized() is prepared.
  //    users can comment next line.
  delay(CZR_INIT_DELAY);
}


bool COZIR::isInitialized()
{
  return (millis() - _initTimeStamp) > CZR_INIT_DELAY;
}


////////////////////////////////////////////////////////////
//
// OPERATING MODE
//
// note: use CZR_COMMAND to minimize power consumption
// CZR_POLLING and CZR_STREAMING use an equally amount
// of power as both sample continuously...
//
bool COZIR::setOperatingMode(uint8_t mode)
{
  if (mode > CZR_POLLING) return false;
  _operatingMode = mode;
  sprintf(_buffer, "K %u", mode);
  _command(_buffer);
  return true;
}


////////////////////////////////////////////////////////////
//
// POLLING MODE
//
// you need to set the polling mode explicitly before
// using these functions. SetOperatingMode(CZR_POLLING);
// this is the default behaviour of this Class but
// not of the sensor!!
//
float COZIR::celsius()
{
  uint16_t rv = _request("T");
  return 0.1 * (rv - 1000.0);
}


float COZIR::humidity()
{
  return 0.1 * _request("H");
}


// UNITS UNKNOWN lux??
float COZIR::light()
{
  return 1.0 * _request("L");
}


uint32_t COZIR::CO2()
{
  return _request("Z");
}


uint16_t COZIR::getPPMFactor()
{
  _ppmFactor = _request(".");
  return _ppmFactor;
}

// CALLIBRATION - USE THESE WITH CARE
// use these only in polling mode (on the Arduino)

// FineTuneZeroPoint()
// a reading of v1 will be reported as v2
// sort of mapping / offset
// check datasheet for detailed description
uint16_t COZIR::fineTuneZeroPoint(uint16_t v1, uint16_t v2)
{
  sprintf(_buffer, "F %u %u", v1, v2);
  return _request(_buffer);
}


// mostly the default calibrator
uint16_t COZIR::calibrateFreshAir()
{
  return _request("G");
}


uint16_t COZIR::calibrateNitrogen()
{
  return _request("U");
}


uint16_t COZIR::calibrateKnownGas(uint16_t value)
{
  sprintf(_buffer, "X %u", value);
  return _request(_buffer);
}


//uint16_t COZIR::calibrateManual(uint16_t value)
//{
  //sprintf(_buffer, "u %u", value);
  //return _request(_buffer);
//}

//uint16_t COZIR::setSpanCalibrate(uint16_t value)
//{
  //sprintf(_buffer, "S %u", value);
  //return _request(_buffer);
//}

//uint16_t COZIR::getSpanCalibrate()
//{
//  return _request("s");
//}


void COZIR::setDigiFilter(uint8_t value)
{
  sprintf(_buffer, "A %u", value);
  _command(_buffer);
}


uint8_t COZIR::getDigiFilter()
{
  return _request("a");
}


////////////////////////////////////////////////////////////
//
// STREAMING MODE
//
// output fields should be OR-ed
// e.g. SetOutputFields(CZR_HUMIDITY | CZR_RAWTEMP | CZR_RAWCO2);
//
// you need to set the STREAMING mode explicitly
// SetOperatingMode(CZR_STREAMING);
//
// in STREAMING mode you must parse the output of serial yourself.
// stream looks like [space field space value]*  \n
//
// - find separator ('\n')
// - read until next separator ('\n') in a buffer,
// - parse buffer [field space value]
//
void COZIR::setOutputFields(uint16_t fields)
{
  _outputFields = fields;
  sprintf(_buffer, "M %u", fields);
  _command(_buffer);
}


bool COZIR::inOutputFields(uint16_t field)
{
  return (_outputFields & field) == field;
}


// WARNING:
// After a call to GetRecentFields() you must read the serial port yourself as
// the internal buffer of this Class cannot handle the possible large output.
// It can be over 100 bytes long lines!
void COZIR::getRecentFields()
{
  _command("Q");
}

////////////////////////////////////////////////////////////
//
// EEPROM CALLS - USE WITH CARE
//
void COZIR::setAutoCalibrationPreload(uint16_t value)
{
  _setEEPROM2(CZR_ACINITHI, value);
}

uint16_t COZIR::getAutoCalibrationPreload()
{
  return _getEEPROM2(CZR_ACINITHI);
}

void COZIR::setAutoCalibrationInterval(uint16_t value)
{
  _setEEPROM2(CZR_ACHI, value);
}

uint16_t COZIR::getAutoCalibrationInterval()
{
  return _getEEPROM2(CZR_ACHI);
}

void COZIR::setAutoCalibrationOn()
{
  _setEEPROM(CZR_ACONOFF, 1);
}

void COZIR::setAutoCalibrationOff()
{
  _setEEPROM(CZR_ACONOFF, 0);
}

bool COZIR::getAutoCalibration()
{
  return _getEEPROM(CZR_ACONOFF);
}

void COZIR::setAutoCalibrationBackgroundConcentration(uint16_t value)
{
  _setEEPROM2(CZR_ACPPMHI, value);
}

uint16_t COZIR::getAutoCalibrationBackgroundConcentration()
{
  return _getEEPROM2(CZR_ACPPMHI);
}

void COZIR::setAmbientConcentration(uint16_t value)
{
  _setEEPROM2(CZR_AMBHI, value);
}

uint16_t COZIR::getAmbientConcentration()
{
  return _getEEPROM2(CZR_AMBHI);
}

void COZIR::setBufferClearTime(uint16_t value)
{
  _setEEPROM2(CZR_BCHI, value);
}

uint16_t COZIR::getBufferClearTime()
{
  return _getEEPROM2(CZR_BCHI);
}


/*
// TODO first verify if single functions work.

void COZIR::setEEPROMFactoryReset()
{
  setAutoCalibrationPreload(0x57C0);
  setAutoCalibrationInterval(0x8E80);
  setAutoCalibrationOff();
  setAutoCalibrationBackgroundConcentration(0x01C2);
  setAmbientConcentration(0x01C2);
  setBufferClearTime(0x0008);
}
*/



////////////////////////////////////////////////////////////
//
// COMMAND MODE
//
// read serial yourself -
//
// TODO Page 5:  Mode 0 Command Mode
// This is primarily intended for use when extracting larger chunks
// of information from the sensor (for example using the Y and * commands).
// In this mode, the sensor is stopped waiting for commands.
//
void COZIR::getVersionSerial()
{
  // override modes to prevent interference in output
  setOperatingMode(CZR_COMMAND);
  _command("Y");
}


void COZIR::getConfiguration()
{
  // override modes to prevent interference in output
  setOperatingMode(CZR_COMMAND);
  _command("*");
}


/////////////////////////////////////////////////////////
//
// PRIVATE
//
void COZIR::_command(const char* str)
{
  _ser->print(str);
  _ser->print("\r\n");
}


uint32_t COZIR::_request(const char* str)
{
  _command(str);

  // read the answer from serial.
  // TODO: PROPER TIMEOUT CODE.
  // - might be a big delay
  // - what is longest answer possible? CZR_REQUEST_TIMEOUT?
  uint8_t idx = 0;
  uint32_t start = millis();
  while (millis() - start < CZR_REQUEST_TIMEOUT)
  {
    if (_ser->available())
    {
      char c = _ser->read();
      if (c == '\n') break;
      _buffer[idx++] = c;
      _buffer[idx] = '\0';
    }
  }
  // Serial.print("buffer: ");
  // Serial.println(_buffer);
  uint32_t rv = 0;
  // default for PPM is different.
  if (str[0] == '.') rv = 1;
  // do we got the requested field?
  if (strchr(_buffer, str[0]) && (idx > 2))
  {
    rv = atol(&_buffer[2]);
  }
  return rv;
}


void COZIR::_setEEPROM(uint8_t address, uint8_t value)
{
  if (address > CZR_BCLO) return;
  sprintf(_buffer, "P %u %u", address, value);
  _command(_buffer);
}


uint8_t COZIR::_getEEPROM(uint8_t address)
{
  sprintf(_buffer, "p %u", address);
  return _request(_buffer);
}


void COZIR::_setEEPROM2(uint8_t address, uint16_t value)
{
  if (address > CZR_BCLO) return;
  sprintf(_buffer, "P %u %u", address, value >> 8);
  _command(_buffer);
  sprintf(_buffer, "P %u %u", address + 1, value & 0xFF);
  _command(_buffer);
}


uint16_t COZIR::_getEEPROM2(uint8_t address)
{
  sprintf(_buffer, "p %u", address);
  uint16_t val = _request(_buffer) << 8;
  sprintf(_buffer, "p %u", address + 1);
  return val + _request(_buffer);
}



////////////////////////////////////////////////////////////////////////////////
//
// C0ZIRParser
//
C0ZIRParser::C0ZIRParser()
{
  init();
}


void C0ZIRParser::init()
{
  _light              = 0;
  _humidity           = 0;
  _LED_FILT           = 0;
  _LED_RAW            = 0;
  _LED_MAX            = 0;
  _zeroPoint          = 0;
  _temperature_RAW    = 0;
  _temperature_FILT   = 0;
  _LED_signal_FILT    = 0;
  _LED_signal_RAW     = 0;
  _temperature_Sensor = 0;
  _CO2_FILT           = 0;
  _CO2_RAW            = 0;
  _samples            = 0;
  _PPM                = 1;  // Note default one
  _value              = 0;
  _field              = 0;
}


uint8_t C0ZIRParser::nextChar(char c)
{
  static bool skipLine = false;
  uint8_t rv = 0;

  //  SKIP * and Y until next return.
  //  as output of these two commands not handled by this parser
  if ((c == '*') || (c == 'Y')) skipLine = true;
  if (c == '\n') skipLine = false;
  if (skipLine) return 0;

  //  TODO investigate
  //  if the last char is more than 2..5 ms ago (9600 baud ~ 1 char/ms)
  //  it probably needs to sync with the stream again.
  //  but it depends on how calling process behaves.
  //  - need for uint32_t _lastChar time stamp?

  switch(c)
  {
    case '0' ... '9':
      _value *= 10;
      _value += (c - '0');
      break;
    //  major responses to catch
    case 'z':
    case 'Z':
    case 'L':
    case 'T':
    case 'H':
    //  all other known responses, starting a new field
    case 'X':
    case '.':
    case 'Y':     // skipped
    case '*':     // skipped
    case 'Q':
    case 'F':
    case 'G':
    case 'M':
    case 'K':    // mode
    case 'A':
    case 'a':
    case 'P':
    case 'p':
    case 'S':
    case 's':
    case 'U':
    case 'u':
    //  new line triggers store() to have results available faster.
    //  saves ~500 millis() for the last FIELD
    case '\n':
      rv = store();
      _field = c;
      _value = 0;
      break;

    //  drop fields of Y, and * command.
    //  reset parsing on separators of Y and * commands
    case ':':
    case ',':
      _field = 0;
      _value = 0;
      break;

    case ' ':    //  known separator
    case '\r':   //  known return
      break;
    default:     //  catch all unknown characters, including glitches.
      break;
  }
  return rv;
}


float C0ZIRParser::celsius()
{
  return  0.1 * (_temperature_FILT - 1000.0);
}


//////////////////////////////////
//
//  PRIVATE
//
uint8_t C0ZIRParser::store()
{
  switch(_field)
  {
    //  LIGHT related
    case 'L':
      _light = _value;
      return _field;
    case 'D':
      _LED_FILT = _value;
      return _field;
    case 'd':
      _LED_RAW = _value;
      return _field;
    case 'l':
      _LED_MAX = _value;
      return _field;
    case 'o':
      _LED_signal_FILT = _value;
      return _field;
    case 'O':
      _LED_signal_RAW = _value;
      return _field;

    // TEMPERATURE & HUMIDITY
    case 'V':
      _temperature_RAW = _value;
      return _field;
    case 'v':
      _temperature_Sensor = _value;
      return _field;
    case 'T':
      _temperature_FILT = _value;
      return _field;
    case 'H':
      _humidity = _value;
      return _field;

    //  CO2 related
    case 'z':
      _CO2_RAW = _value;
      return _field;
    case 'Z':
      _CO2_FILT = _value;
      return _field;

    //  OTHER
    case 'h':
      _zeroPoint = _value;
      return _field;
    case 'a':
      _samples = _value;
      return _field;
    case '.':
      _PPM = _value;
      return _field;
  }
  return 0;
}


// -- END OF FILE --



cozir.h
#pragma once
//
//    FILE: Cozir.h
// VERSION: 0.3.5
// PURPOSE: library for COZIR range of sensors for Arduino
//          Polling Mode + stream parser
//     URL: https://github.com/RobTillaart/Cozir
//          http://forum.arduino.cc/index.php?topic=91467.0
//
// READ DATASHEET BEFORE USE OF THIS LIB !
//


#include "Arduino.h"


#define COZIR_LIB_VERSION           (F("0.3.5"))


// OUTPUT FIELDS
// See datasheet for details.
// These defines can be OR-ed for the SetOutputFields command
//
#define CZR_UNKNOWN_2               0x8000     // returns  P 00128  ? 
#define CZR_UNKNOWN_1               0x4000     // returns  E 00016  ?
#define CZR_LIGHT                   0x2000
#define CZR_HUMIDITY                0x1000
#define CZR_FILTLED                 0x0800
#define CZR_RAWLED                  0x0400
#define CZR_MAXLED                  0x0200
#define CZR_ZEROPOINT               0x0100
#define CZR_RAWTEMP                 0x0080
#define CZR_FILTTEMP                0x0040
#define CZR_FILTLEDSIGNAL           0x0020
#define CZR_RAWLEDSIGNAL            0x0010
#define CZR_SENSTEMP                0x0008
#define CZR_FILTCO2                 0x0004
#define CZR_RAWCO2                  0x0002

#define CZR_NONE                    0x0001


// factory default 
#define CZR_DEFAULT                 (CZR_FILTCO2 | CZR_RAWCO2)
// easy default setting for streaming
#define CZR_HTC                     (CZR_HUMIDITY | CZR_RAWTEMP | CZR_RAWCO2)
// not in datasheet for debug only
#define CZR_ALL                     0x3FFE


// OPERATING MODES
#define CZR_COMMAND                 0x00
#define CZR_STREAMING               0x01
#define CZR_POLLING                 0x02


class COZIR
{
public:
  COZIR(Stream * str);
  void     init();      // sets operatingMode to CZR_POLLING
  bool     isInitialized();

  // warning: CZR_STREAMING is experimental, minimal tested.
  bool     setOperatingMode(uint8_t mode);
  uint8_t  getOperatingMode() { return _operatingMode; };


  //  POLLING MODE
  float    celsius();
  float    fahrenheit() { return (celsius() * 1.8) + 32; };
  float    kelvin()     { return celsius() + 273.15; };
  float    humidity();
  float    light();
  uint32_t CO2();
  uint16_t getPPMFactor();   // P14 . command  return 1, 10 or 100


  //  CALIBRATION
  //  read datasheet before use
  uint16_t fineTuneZeroPoint(uint16_t v1, uint16_t v2);
  uint16_t calibrateFreshAir();
  uint16_t calibrateNitrogen();
  uint16_t calibrateKnownGas(uint16_t value);

  //  WARNING: following 3 functions are NOT RECOMMENDED,
  //           read datasheet before use
  //  uint16_t calibrateManual(uint16_t value);
  //  uint16_t setSpanCalibrate(uint16_t value);
  //  uint16_t getSpanCalibrate();


  //  DIGIFILTER, use with care, read datasheet before use
  //     32 = good default value,
  //      1 = fast (noisy and responsive to fast changes)
  //    255 = slow (smoothed to the max)
  //      0 = special. details see datasheet
  void     setDigiFilter(uint8_t value);
  uint8_t  getDigiFilter();


  //  STREAMING MODE
  void     setOutputFields(uint16_t fields);
  uint16_t getOutputFields() { return _outputFields; };
  bool     inOutputFields(uint16_t field);
  void     clearOutputFields() { setOutputFields(CZR_NONE); };
  //  WARNING:
  //  After a call to GetRecentFields() you must read the serial port yourself as
  //  the internal buffer of this Class cannot handle the possible large output.
  //  Answers can be over 100 bytes long!
  void     getRecentFields();


  //  EEPROM
  //  not all sensors support these.
  void     setAutoCalibrationPreload(uint16_t value);
  uint16_t getAutoCalibrationPreload();

  void     setAutoCalibrationInterval(uint16_t value);
  uint16_t getAutoCalibrationInterval();

  void     setAutoCalibrationOn();
  void     setAutoCalibrationOff();
  bool     getAutoCalibration();

  void     setAutoCalibrationBackgroundConcentration(uint16_t value);
  uint16_t getAutoCalibrationBackgroundConcentration();

  void     setAmbientConcentration(uint16_t value);
  uint16_t getAmbientConcentration();

  void     setBufferClearTime(uint16_t value);
  uint16_t getBufferClearTime();

  // TODO test EEPROM function first.
  // void setEEPROMFactoryReset();


  //  META INFORMATION
  //  library does not parse the output (yet)
  void     getVersionSerial();
  void     getConfiguration();


  ///////////////////////////////////////////////
  //
  // SEMI PRIVATE FOR UNIT TESTING THEM
  //
  void     _setEEPROM(uint8_t address, uint8_t value);
  uint8_t  _getEEPROM(uint8_t address);
  void     _setEEPROM2(uint8_t address, uint16_t value);
  uint16_t _getEEPROM2(uint8_t address);


private:
  Stream * _ser;
  char     _buffer[20];

  uint32_t _initTimeStamp = 0;
  uint16_t _ppmFactor     = 1;

  uint8_t  _operatingMode = CZR_STREAMING;
  uint16_t _outputFields  = CZR_NONE;

  void     _command(const char* str);
  uint32_t _request(const char* str);
};



////////////////////////////////////////////////////////////////////////////////
//
//  C0ZIRParser
//
//  used to parse the stream from a COZIR CO2 sensor.
//  Note: one can comment fields / code not used to minimize footprint.
//
class C0ZIRParser
{
public:
  C0ZIRParser();

  void init();

  // returns field char if a field is completed, 0 otherwise.
  uint8_t nextChar(char c);

  //  FETCH LAST READ VALUES
  float    celsius();
  float    fahrenheit()    { return (celsius() * 1.8) + 32; };
  float    kelvin()        { return celsius() + 273.15; };
  float    humidity()      { return 0.1 * _humidity; };

  uint16_t light()         { return _light; };
  uint16_t ledFilt()       { return _LED_FILT; };
  uint16_t ledRaw()        { return _LED_RAW; };
  uint16_t ledMax()        { return _LED_MAX; };
  uint16_t ledSignalFilt() { return _LED_signal_FILT; };
  uint16_t ledSignalRaw()  { return _LED_signal_RAW; };

  uint16_t zeroPoint()     { return _zeroPoint; };
  uint16_t tempFilt()      { return _temperature_FILT; };
  uint16_t tempRaw()       { return _temperature_RAW; };
  uint16_t tempSensor()    { return _temperature_Sensor; };

  uint16_t CO2()           { return _CO2_FILT; };
  uint16_t CO2Raw()        { return _CO2_RAW; };

  uint16_t samples()       { return _samples; };
  uint16_t getPPMFactor()  { return _PPM; }


private:
  //       FIELD                    ID character
  uint16_t _light;              //  L
  uint16_t _humidity;           //  H
  uint16_t _LED_FILT;           //  D
  uint16_t _LED_RAW;            //  d
  uint16_t _LED_MAX;            //  l    // el not one
  uint16_t _zeroPoint;          //  h
  uint16_t _temperature_RAW;    //  V
  uint16_t _temperature_FILT;   //  T
  uint16_t _LED_signal_FILT;    //  o    // oo not zero
  uint16_t _LED_signal_RAW;     //  O    // oo not zero
  uint16_t _temperature_Sensor; //  v
  uint16_t _CO2_FILT;           //  Z
  uint16_t _CO2_RAW;            //  z

  // not output fields sec but useful.
  uint16_t _samples;            //  a
  uint16_t _PPM;                //  .    // point


  // parsing helpers
  uint32_t _value;
  uint8_t  _field;

  // returns field char if a field is completed, 0 otherwise.
  uint8_t store();
};


// -- END OF FILE --

